src/metrics/mod.rs

```diff
@@ -67,6 +78,64 @@ metrics! {
     namespace: "docsrs",
 }
 
+#[derive(Debug, Default)]
+pub(crate) struct RecentlyAccessedReleases {
+    krates: DashMap<i32, Instant>,
+    versions: DashMap<i32, Instant>,
+    platforms: DashMap<(i32, TargetAtom), Instant>,
+}
+
+impl RecentlyAccessedReleases {
+    pub(crate) fn new() -> Self {
+        Self::default()
+    }
+
+    pub(crate) fn record(&self, krate: i32, version: i32, target: &str) {
+        self.krates.insert(krate, Instant::now());
+        self.versions.insert(version, Instant::now());
+        self.platforms
+            .insert((version, TargetAtom::from(target)), Instant::now());
+    }
+
+    pub(crate) fn gather(&self, metrics: &Metrics) {
+        fn inner<K: std::hash::Hash + Eq>(map: &DashMap<K, Instant>, metric: &IntGaugeVec) {
+            let mut hour_count = 0;
+            let mut half_hour_count = 0;
+            let mut five_minute_count = 0;
+            map.retain(|_, instant| {
+                let elapsed = instant.elapsed();
+                if elapsed > Duration::from_secs(60 * 60) {
+                    return false;
+                }
+                hour_count += 1;
+                if elapsed > Duration::from_secs(30 * 60) {
+                    return true;
+                }
+                half_hour_count += 1;
+                if elapsed > Duration::from_secs(5 * 60) {
+                    return true;
+                }
+                five_minute_count += 1;
+                true
+            });
```

Comment:
This is really confusing. What is it doing? Why does it return `false` after an hour? Why does it return early if _more_ than the time elapsed has passed - couldn't it skip recording an interval?

Reply:
It goes through from longest to shortest interval, checking if the value is within the interval, short-circuiting and exiting if it's not, otherwise incrementing that counter and moving on to check the next interval. The first check returns `false` to remove the entry from the map, the others return `true` to keep it for future checks.

Reply:
I can probably rewrite it to be more obvious, and just rely on the optimizer to reorder the duplicate checks that that would need.